### PR TITLE
Remove pkg resources

### LINKED ISSUE #1970: Opentelemetry instrumentation uses deprecated pkg_resources.
**Describe your environment** 
Python 3.11
Setuptools >= 67.5.0
Windows 11

**Steps to reproduce**

Any operation that imports opentelemetry\instrumentation\dependencies.py

**What is the expected behavior?**

No deprecation warnings

**What is the actual behavior?**
What did you see instead?

`C:\Users\jenielse\Miniconda3\envs\myenv\Lib\site-packages\opentelemetry\instrumentation\dependencies.py:4: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html`

**Additional context**

This was removed from opentelemetry sdk/api in https://github.com/open-telemetry/opentelemetry-python/pull/3047

In addition to being deprecated pkg_resources is known to be problematic since it indexes all packages on start up adding significant overhead to the package import. Normally this should be replaced by importlib.resources or for older python version support importlib_resources  

As a bonus this will also remove the runtime dependency on setuptools

### LINKED ISSUE #2180: Remove pkg_resource and replace for importlib.metadata
**Describe your environment**

Docker environment
Image so: ubuntu:20.04
Python version: 3.9

Otel dependencies:

opentelemetry-api = "^1.14.0"
opentelemetry-semantic-conventions = "0.41b0"  # At the moment there are no releases without beta
opentelemetry-instrumentation-urllib3 = "0.41b0"  # At the moment there are no releases without beta
opentelemetry-instrumentation-httpx = "0.41b0" # At the moment there are no releases without beta

**Steps to reproduce**

Instrument any app and run it, it will throws a warning deprecation message about pkg_resource.

I created a CM of spans with something like this:

``` python
from opentelemetry import trace
@contextlib.contextmanager
def create_span(tracer: trace.Tracer) -> Generator:
    "Start the span as current span."
    name = "name"
    _kind = "kind"
    with tracer.start_as_current_span(
        name,
        kind=_kind,
        record_exception=False,
    ) as current_span:
        yield current_span
```

**What is the expected behavior?**

Otel doesnt throws a deprecationWarning

**What is the actual behavior?**

The output is a deprecation warning:

`DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html`

**Additional context**

[Use of pkg_resources is deprecated](https://setuptools.pypa.io/en/latest/pkg_resources.html)

### LINKED ISSUE #2843: Importing `opentelemetry.instrumentation.aws_lambda` module is slow because of `pkg_resources`
### Describe your environment

Environment: `AWS Lambda`
Python version: `Python 3.9`
Package version: `0.47b0`

### What happened?

Initialization of the OTEL Python Lambda layer takes significant amount of time in the AWS Lambda environment because of module imports. When I have analyzed import timings, I have noticed that importing `opentelemetry.instrumentation.aws_lambda` is slow, because of 
`opentelemetry.instrumentation.instrumentor` > `opentelemetry.instrumentation.dependencies` > `pkg_resources` imports as shown below.

Can we make importing `opentelemetry.instrumentation.dependencies` in the `instrumentor.py` lazy and optional (ignore when `skip_dep_check` arg is `True`) here: https://github.com/open-telemetry/opentelemetry-python-contrib/blob/98958b6f1ef09d332e13ea1d7bc308f8e9edd351/opentelemetry-instrumentation/src/opentelemetry/instrumentation/instrumentor.py#L84 So we can configure/optimize it in the AWS Lambda Python layer side.

Portion of the module import timings related to `opentelemetry.instrumentation.aws_lambda` module import in a Lambda function with `512 MB` memory. 
```
import time:      3143 |       3143 |         pkgutil
import time:       437 |        437 |               xml
import time:       356 |        793 |             xml.parsers
import time:       700 |        700 |             pyexpat
import time:       326 |       1817 |           xml.parsers.expat
import time:      1521 |       3338 |         plistlib
import time:      1121 |       1121 |           packaging
import time:       677 |        677 |                     sysconfig
import time:      1395 |       1395 |                       packaging._elffile
import time:      2261 |       3655 |                     packaging._manylinux
import time:      1225 |       1225 |                     packaging._musllinux
import time:      3802 |       9357 |                   packaging.tags
import time:       691 |        691 |                     packaging._structures
import time:      5171 |       5861 |                   packaging.version
import time:      2460 |      17677 |                 packaging.utils
import time:     11271 |      28948 |               packaging.specifiers
import time:      7844 |      36791 |             packaging._tokenizer
import time:      2756 |      39546 |           packaging._parser
import time:      3292 |      43958 |         packaging.markers
import time:       974 |        974 |         packaging.requirements
import time:       946 |        946 |           jaraco
import time:      4772 |       4772 |                 more_itertools.recipes
import time:     22981 |      27752 |               more_itertools.more
import time:       472 |      28223 |             more_itertools
import time:      2285 |      30508 |           jaraco.functools
import time:      1831 |       1831 |             backports
import time:       312 |        312 |                 backports.tarfile.compat
import time:      1558 |       1870 |               backports.tarfile.compat.py38
import time:     21746 |      23615 |             backports.tarfile
import time:      2868 |      28313 |           jaraco.context
import time:      4887 |      64653 |         jaraco.text
import time:      4140 |       4140 |           platformdirs.api
import time:       357 |        357 |           platformdirs.version
import time:      2635 |       2635 |             configparser
import time:      1841 |       4476 |           platformdirs.unix
import time:      4259 |      13230 |         platformdirs
import time:     66019 |     195313 |       pkg_resources
import time:      2064 |     197376 |     opentelemetry.instrumentation.dependencies
import time:      1511 |     248706 |   opentelemetry.instrumentation.instrumentor
import time:      2923 |     252130 | opentelemetry.instrumentation.aws_lambda
```

### Steps to Reproduce

- Set `PYTHONPROFILEIMPORTTIME` env var to `true` in the `otel-instrument` bash script
- Import `opentelemetry.instrumentation.aws_lambda` module

### Expected Result

-

### Actual Result

-

### Additional context

_No response_

### Would you like to implement a fix?

None

### PR DESCRIPTION
**Pull request by xrmx**
# Description

This incorporates all previous work from @Rodrigo-Novas and @ocelotl and makes ci happy.

Fixes #2180
Fixes #1970
Fixes #2843

Closes #1973 #2124 #2181 #2854

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-fastapi/tests/test_fastapi_instrumentation.py [hunk 1]
```diff
@@ -1048,30 +1048,18 @@ class TestAutoInstrumentation(TestBaseAutoFastAPI):
...
-        self.assertEqual(ep.dist.key, "opentelemetry-instrumentation-fastapi")
-        self.assertEqual(
-            ep.module_name, "opentelemetry.instrumentation.fastapi"
-        )
-        self.assertEqual(ep.attrs, ("FastAPIInstrumentor",))
+        (ep,) = entry_points(group="opentelemetry_instrumentor")
         self.assertEqual(ep.name, "fastapi")
-        self.assertIsNone(next(eps, None))
 
+    """ FIXME: get_distribution is gone
```
Comment (xrmx) [inline instrumentation/opentelemetry-instrumentation-fastapi/tests/test_fastapi_instrumentation.py:1054]: This needs to be sorted out before making this ready

File: instrumentation/opentelemetry-instrumentation-wsgi/tests/__init__.py [hunk 2]
```diff
@@ -1,23 +0,0 @@
-# Copyright The OpenTelemetry Authors
```
Comment (xrmx) [inline instrumentation/opentelemetry-instrumentation-wsgi/tests/__init__.py:1]: YOLO

File: opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py [hunk 3]
```diff
@@ -27,13 +26,36 @@
...
+    distributions,
+    entry_points,
+)
 
 _logger = getLogger(__name__)
 
 
+class _EntryPointDistFinder:
+    @cached_property
+    def _mapping(self):
```
Comment (xrmx) [inline opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:40]: In tests we don't pass here so two options: 1) is dead code 2) we need to add a test

Comment (xrmx) [reply opentelemetry-instrumentation/src/opentelemetry/instrumentation/auto_instrumentation/_load.py:40]: Add tests, still not sure this is required :)

**Comment by mlavin** [comment]
Tested out locally and it works for me. My benchmarks show a ~100-200ms improvement in the import/start time when manually configuring and ~400-500ms when using the auto-instrumentation. Thank you to everyone who contributed towards this change and I can't wait to see this land 🙌

File: CHANGELOG.md [hunk 4]
```diff
@@ -25,12 +25,16 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `opentelemetry-instrumentation-aiokafka` Wrap `AIOKafkaConsumer.getone()` instead of `AIOKafkaConsumer.__anext__`
   ([#2874](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2874))
 
+### Breaking changes
+
+- Deprecation of pkg_resource in favor of importlib.metadata
+  ([#2181](https://github.com/open-telemetry/opentelemetry-python-contrib/issues/2181))
```
Comment (emdneto) [inline CHANGELOG.md:31]:
```suggestion
  ([#2871](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2871))
```

File: instrumentation/opentelemetry-instrumentation-fastapi/tests/test_fastapi_instrumentation.py [hunk 5]
```diff
@@ -1028,16 +1031,24 @@ def client_response_hook(send_span, scope, message):
...
-def get_distribution_with_fastapi(*args, **kwargs):
-    dist = args[0]
-    if dist == "fastapi~=0.58":
-        # Value does not matter. Only whether an exception is thrown
-        return None
-    raise DistributionNotFound()
+def mock_version_with_fastapi(*args, **kwargs):
+    req_name = args[0]
+    if req_name == "fastapi":
+        # TODO: Value now matters
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-fastapi/tests/test_fastapi_instrumentation.py:1037]: What's the TODO here? It's something we should fix later?

File: instrumentation/opentelemetry-instrumentation-test/tests/__init__.py [hunk 6]
```diff
@@ -1,23 +0,0 @@
...
-import pkg_resources
-
-# IMPORTANT: Only the TEst module needs this because it is always the first
-# package that uses the `{rootdir}/*/tests/` path and gets installed by
-# `eachdist.py` and according to `eachdist.ini`.
-
-# Naming the tests module as a namespace package ensures that
-# relative imports will resolve properly for subsequent test packages,
-# as it enables searching for a composite of multiple test modules.
-pkg_resources.declare_namespace(__name__)
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-test/tests/__init__.py:23]: Just to make sure, we don't need to explicitly declare namespace packages anymore right?

**Review by lzchen — APPROVED** [review]

**Review by tammy-baylis-swi — APPROVED** [review]

File: instrumentation/opentelemetry-instrumentation-fastapi/tests/test_fastapi_instrumentation.py (hunk 5 above)
Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-fastapi/tests/test_fastapi_instrumentation.py:1037]: This is from Jeremy, I think it means it should be updated if the one in pyproject changes.

File: instrumentation/opentelemetry-instrumentation-test/tests/__init__.py (hunk 6 above)
Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-test/tests/__init__.py:23]: Not sure but tests are green :)

**Review by emdneto — APPROVED** [review]

File: instrumentation/opentelemetry-instrumentation-test/tests/__init__.py (hunk 6 above)
Comment (emdneto) [reply instrumentation/opentelemetry-instrumentation-test/tests/__init__.py:23]: eachdist.py isn't breaking as well

**Comment by moserke** [comment]
Do we know when this will make it into a release?  Looks like there has not been a release since August, will one be coming soon?  I see this in the CHANGELOG.md.  It's blocking us from being able to go to python 3.12, so mostly just curious on timing.

**Comment by tammy-baylis-swi** [comment]
> Do we know when this will make it into a release? Looks like there has not been a release since August, will one be coming soon?

Yes, this is now available as part of 1.28.1/0.49b1.

**Pull request merged**